mediaengine.go

```diff
@@ -561,18 +561,24 @@ func (m *MediaEngine) getRTPParametersByKind(typ RTPCodecType, directions []RTPT
 
 	m.mu.RLock()
 	defer m.mu.RUnlock()
-	if m.negotiatedVideo && typ == RTPCodecTypeVideo ||
-		m.negotiatedAudio && typ == RTPCodecTypeAudio {
-		for id, e := range m.negotiatedHeaderExtensions {
-			if haveRTPTransceiverDirectionIntersection(e.allowedDirections, directions) && (e.isAudio && typ == RTPCodecTypeAudio || e.isVideo && typ == RTPCodecTypeVideo) {
-				headerExtensions = append(headerExtensions, RTPHeaderExtensionParameter{ID: id, URI: e.uri})
-			}
-		}
+
+	var mediaHeaderExtensions map[int]mediaEngineHeaderExtension
+	if (m.negotiatedVideo && typ == RTPCodecTypeVideo) || (m.negotiatedAudio && typ == RTPCodecTypeAudio) {
+		mediaHeaderExtensions = m.negotiatedHeaderExtensions
 	} else {
+		mediaHeaderExtensions = make(map[int]mediaEngineHeaderExtension, len(m.headerExtensions))
 		for id, e := range m.headerExtensions {
-			if haveRTPTransceiverDirectionIntersection(e.allowedDirections, directions) && (e.isAudio && typ == RTPCodecTypeAudio || e.isVideo && typ == RTPCodecTypeVideo) {
-				headerExtensions = append(headerExtensions, RTPHeaderExtensionParameter{ID: id + 1, URI: e.uri})
-			}
+			mediaHeaderExtensions[id+1] = e
+		}
+	}
+
+	for id, e := range mediaHeaderExtensions {
+		if (typ == RTPCodecTypeVideo && !e.isVideo) || (typ == RTPCodecTypeAudio && !e.isAudio) {
+			continue
+		}
+
+		if haveRTPTransceiverDirectionIntersection(e.allowedDirections, directions) {
+			headerExtensions = append(headerExtensions, RTPHeaderExtensionParameter{ID: id, URI: e.uri})
```

Comment:
I found this section hard to read to reason. So, I refactored it. Hope it is okay

Reply:
please refactor away :) I would love whole PRs that just move code around/make it simpler. The more involved you are in the code base the better.